pkg/api/vm/handler.go

```diff
@@ -626,6 +642,119 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
 		Error()
 }
 
+// cloneVM creates a VM which uses volume cloning from the source VM.
+func (h *vmActionHandler) cloneVM(name string, namespace string, input CloneInput) error {
+	vm, err := h.vmCache.Get(namespace, name)
+	if err != nil {
+		return fmt.Errorf("cannot get vm %s/%s, err: %w", namespace, name, err)
+	}
+	newVM := &kubevirtv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        input.TargetVM,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+			Labels:      vm.Labels,
+		},
+		Spec: *vm.Spec.DeepCopy(),
+	}
+	newVM.Spec.Template.Spec.Hostname = newVM.Name
+	newVM.Spec.Template.ObjectMeta.Labels[builder.LabelKeyVirtualMachineName] = newVM.Name
+
+	newPVCs, secretNameMap, err := h.cloneVolumes(newVM)
+	if err != nil {
+		return fmt.Errorf("clone volumes error for new vm %s/%s, err %w", newVM.Namespace, newVM.Name, err)
+	}
+	newPVCsString, err := json.Marshal(newPVCs)
+	if err != nil {
+		return fmt.Errorf("cannot marshal value %+v, err: %w", newPVCs, err)
+	}
+
+	newVM.ObjectMeta.Annotations[util.AnnotationVolumeClaimTemplates] = string(newPVCsString)
+	if newVM, err = h.vms.Create(newVM); err != nil {
+		return fmt.Errorf("cannot create newVM %+v, err: %w", newVM, err)
+	}
+
+	for oldSecretName, newSecretName := range secretNameMap {
+		secret, err := h.secretCache.Get(namespace, oldSecretName)
+		if err != nil {
+			return fmt.Errorf("cannot get secret %s/%s, err: %w", namespace, oldSecretName, err)
+		}
+
+		newSecret := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      newSecretName,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: newVM.APIVersion,
+						Kind:       newVM.Kind,
+						Name:       newVM.Name,
+						UID:        newVM.UID,
+					},
+				},
+			},
+			Data:       secret.Data,
+			StringData: secret.StringData,
+			Type:       secret.Type,
+		}
+		if _, err = h.secretClient.Create(&newSecret); err != nil {
+			return fmt.Errorf("cannot create a new secret from %s/%s, err: %w", namespace, oldSecretName, err)
+		}
+	}
+	return nil
+}
+
+func (h *vmActionHandler) cloneVolumes(newVM *kubevirtv1.VirtualMachine) ([]corev1.PersistentVolumeClaim, map[string]string, error) {
```

Comment:
just wondering why don't we create the new PVC and secrets directly first?

Reply:
For PVC, we follow the same steps to create PVC for a new VM, so we let the controller handle it. In the future, if we want to change some implementation of creating PVC, we can only update the controller part.

For secret, if we create a secret first, we have to send an update request again to add the owner reference. Currently, for a new VM, we also create VM first and then secrets.

![Screen Shot 2022-06-20 at 1 53 56 PM](https://user-images.githubusercontent.com/4927361/174534413-93e92786-6f28-48ac-94f9-5d79a019e1a8.png)
![Screen Shot 2022-06-20 at 1 54 02 PM](https://user-images.githubusercontent.com/4927361/174534439-3ea48bdc-9713-4cd1-8986-1a1fdba5d8d5.png)